contrib/src/templates/fairing.rs

```diff
@@ -0,0 +1,125 @@
+use super::DEFAULT_TEMPLATE_DIR;
+use super::context::Context;
+use super::engine::Engines;
+
+#[cfg(debug_assertions)]
+use super::watch::TemplateWatcher;
+
+use rocket::{Data, Request, Rocket};
+use rocket::config::ConfigError;
+use rocket::fairing::{Fairing, Info, Kind};
+
+#[cfg(not(debug_assertions))]
+mod context {
+    use std::ops::Deref;
+    use super::Context;
+
+    /// Wraps a Context. With `cfg(debug_assertions)` active, this structure
+    /// additionally provides a method to reload the context at runtime.
+    pub struct ContextManager(Context);
+
+    impl ContextManager {
+        pub fn new(ctxt: Context) -> ContextManager {
+            ContextManager(ctxt)
+        }
+
+        pub fn get<'a>(&'a self) -> impl Deref<Target=Context> + 'a {
+            &self.0
+        }
+    }
+}
+
+#[cfg(debug_assertions)]
+mod context {
+    use std::ops::{Deref, DerefMut};
+    use std::sync::RwLock;
+    use super::{Context, Engines, TemplateWatcher};
+
+    /// Wraps a Context. With `cfg(debug_assertions)` active, this structure
+    /// additionally provides a method to reload the context at runtime.
+    pub struct ContextManager{ context: RwLock<Context>, watcher: Option<TemplateWatcher> }
+
+    impl ContextManager {
+        pub fn new(ctxt: Context) -> ContextManager {
+            let root = ctxt.root.clone();
+            ContextManager {
+                context: RwLock::new(ctxt),
+                watcher: TemplateWatcher::new(root),
+            }
+        }
+
+        pub fn get<'a>(&'a self) -> impl Deref<Target=Context> + 'a {
+            self.context.read().unwrap()
+        }
+
+        fn get_mut<'a>(&'a self) -> impl DerefMut<Target=Context> + 'a {
+            self.context.write().unwrap()
+        }
+
+        pub fn reload_if_needed<F: Fn(&mut Engines)>(&self, custom_callback: F) {
+            if self.watcher.as_ref().map(TemplateWatcher::needs_reload).unwrap_or(false) {
+                warn!("Change detected, reloading templates");
+                let mut ctxt = self.get_mut();
+                match Context::initialize(ctxt.root.clone(), custom_callback) {
+                    Some(new_ctxt) => { *ctxt = new_ctxt; }
+                    None => { warn!("An error occurred while reloading templates. The previous templates will remain active."); }
+                };
+            }
+        }
+    }
+}
+
+pub use self::context::ContextManager;
+
+pub struct TemplateFairing {
+    custom_callback: Box<Fn(&mut Engines) + Send + Sync + 'static>,
```

Comment:
You should create a type alias for the function type instead of using the long type everywhere.

Reply:
What should this be named? TemplateCustomizationCallback? Or maybe just CustomizationCallback?

Reply:
Maybe just `CustomEngineFn`? Something ending in `Fn` would be good.

Reply:
This is all that changes, is it still worth it?
```diff
diff --git a/contrib/src/templates/fairing.rs b/contrib/src/templates/fairing.rs
index ab15dd3..74c6e12 100644
--- a/contrib/src/templates/fairing.rs
+++ b/contrib/src/templates/fairing.rs
@@ -104,12 +104,14 @@ mod context {
 
 pub use self::context::ContextManager;
 
+pub type CustomEngineFn = Fn(&mut Engines) + Send + Sync + 'static;
+
 pub struct TemplateFairing {
-    custom_callback: Box<Fn(&mut Engines) + Send + Sync + 'static>,
+    custom_callback: Box<CustomEngineFn>,
 }
 
 impl TemplateFairing {
-    pub fn new(custom_callback: Box<Fn(&mut Engines) + Send + Sync + 'static>) -> TemplateFairing
+    pub fn new(custom_callback: Box<CustomEngineFn>) -> TemplateFairing
     {
         TemplateFairing { custom_callback }
     }
```

The only other place all the bounds show up is in a trait bound, where the type alias can't be used.